cpp/src/arrow/compute/kernels/util_internal.h

```diff
@@ -30,6 +30,24 @@ namespace arrow {
 namespace compute {
 namespace internal {
 
+// Used in some kernels and testing - not provided by default in MSVC
+// and _USE_MATH_DEFINES is not reliable with unity builds
+#ifndef M_E
+#define M_E 2.71828182845904523536
+#define M_LOG2E 1.44269504088896340736
+#define M_LOG10E 0.434294481903251827651
+#define M_LN2 0.693147180559945309417
+#define M_LN10 2.30258509299404568402
+#define M_PI 3.14159265358979323846
+#define M_PI_2 1.57079632679489661923
+#define M_PI_4 0.785398163397448309616
+#define M_1_PI 0.318309886183790671538
+#define M_2_PI 0.636619772367581343076
+#define M_2_SQRTPI 1.12837916709551257390
+#define M_SQRT2 1.41421356237309504880
+#define M_SQRT1_2 0.707106781186547524401
+#endif
+
```

Comment:
I do not understand the issue with `_USE_MATH_DEFINES` and unity builds. These defines should be guarded with
```
#if defined(_USE_MATH_DEFINES) && !defined(_MATH_DEFINES_DEFINED)
#define _MATH_DEFINES_DEFINED
...
#endif
```
to prevent multiple definitions.

Reply:
It's the opposite: if any translation unit forgets to add that define before including cmath, even if you don't forget, you'll get a build error, since cmath will already have been included.

Reply:
Ok, I see. You are guarding all the defines based on the existence of `M_E`. I think it would be "safer" to guard each macro independently. Safer in the sense that if at some point `M_E` is defined explicitly in another translation unit, then the remaining macros will be undefined.